### ISSUE TITLE
Signing a windows application in Mac Os using a private key located in Yubikey (Fips 140)

### ISSUE THREAD

**Issue by pefka** [electron-userland/electron-builder#7895]
* **Electron-Builder Version**: 23.3.3

* **Node Version**:  16.8

* **Electron Version**: 21.3.1
* **Electron Type (current, beta, nightly)**: current

* **Target**:  Windows

Hello everyone, last year I built and signed a Windows application from under Mac Os, with the usual config, where I specified:
`        win: {
          "target": [
            {
              "target": "nsis",
              "arch": [
                "x64",
                "ia32"
              ]
            }
          ]
          ,
          icon: "build/1024-logo-windows.png",
          signingHashAlgorithms: ['sha256'],
          certificateFile: "certificate/output.pfx",
          certificatePassword: process.env.WIN_CER_PASS
        }`

Now due to new rules June 2023. I bought a Yubikey FIPS 140 usb hardware token. I passed verification and received files like: root certificate, server certificate, etc. from Comodo.

**But I can’t export the private key from Yubikey and just to replace two lines in my config (certificateFile: && certificatePassword:).**

Somehow, the electron builder should see that I have a connected flash drive, and that it should look for the private key in it. How to do it? 
Which lines of the config should I change, what is the general logic of action? 
Please help anyone who has encountered this.

**Comment by mmaietta** [comment]
I'm not familiar with the new ruleset, but there is a thread on implementing the new signing process. Have you taken a look at? https://github.com/electron-userland/electron-builder/issues/7605

**Comment by perry-mitchell** [comment]
@mmaietta That thread is for signing windows executables _on_ Windows, so far as I can tell, not on Mac. That, and they were able to sign using a CI, which I'm not sure how we can do given that we can't export or share the Yubikey's private key.

@pefka Did you solve this by any chance? I'm facing the same issue.. I'm building locally on Mac, as the CI process is dead so far as I can tell.. and I want to sign my windows binaries.

**Comment by pefka** [comment]
@perry-mitchell , 
Yes. I started using a windows computer for this. It's very messy, but it works.

1) I use the old scheme to create a binary from mac os (because I write code on Mac):
vue-cli-service electron:build --win

in the config:
sha256 left the line, commented out the rest

```
 signingHashAlgorithms: ['sha256'],
 // certificateFile: "certificate/output.pfx",
 // certificatePassword: process.env.WIN_CER_PASS,
 // certificateSubjectName: "xxx SOFTWARE xxx"
```

2) on windows I installed this filthy yubikey, having first installed the program for it there. Plus! It turns out that for this program it was necessary to install some kind of general application (something like a service, a daemon), since windows does not always detect yubikey as needed. (so 2 apps from web yubikey need to be installed)

3) wrote to yubikey support, did not receive a normal answer (never buy yubikey for yourself, this is a disgusting company)

4) using signtool I sign my application in Windows with an inserted flash drive (yubikey), a window appears in Windows with entering a PIN code for yubikey

5) the app is signed. I recalculate its hash, upload the signed program to the server and change the hash to a new one in the auto-updater

This way it just works for me. If you need terminal commands in Windows, from points 4 and 5, I can send them. They are on another computer.

**Comment by perry-mitchell** [comment]
Thanks for the detailed response @pefka! 

It's a pity that there seems to be no integration for hardware signing in electron builder, and that we need to resort to _both_ using Windows as well as "automating" it ourselves.

I wonder if I could actually automate this using something like [windows in docker](https://github.com/dockur/windows), but that may be far fetched.

If I figure out a more automatic way of doing this I'll share it here as well. Cheers

**Comment by serg06** [comment]
Hello, I just ran into this exact same issue. Here's what I did to fix it.

## Background

Electron-builder uses `signtool` to sign your app on Windows. If we can get a working Signtool command, and we can patch Electron Builder to use our own Signtool command, we're good.

### Getting the Signtool command working

1. I set up the Windows Yubikey software and driver.
2. I imported the certificatesa. They gave me 4 certificates: Server, intermediate 1, intermediate 2, and root. I put each one in its own text file ending with ".crt".
a. I imported the server.crt into the Yubikey GUI: https://help.comodosslstore.com/support/solutions/articles/22000281501-yubikey-5-fips-hsm-certificate-installation
b. I imported the 3 other certs using the command line: `ykman piv certificates import 82 root.pem`, `ykman piv certificates import 83 intermediate1.pem`, `ykman piv certificates import 84 intermediate2.pem` (Thanks ChatGPT)
3. I got the certificate's sha1: I double-clicked server.crt, clicked Details, and scrolled down to Thumbprint.
4. I created a sample unsigned Hello World.exe file with gcc.
5. I tried signing it with the following command: `sign /debug /sha1 "THUMBPRINT-GOES-HERE" /fd sha256 /t http://timestamp.digicert.com sample.exe`
6. It failed and said it doesn't see my cert.
a. I double-clicked all four of my certs and clicked install.
b. I restarted my PC
c. I left it alone for a couple of days
7. I tried the signing command again and it worked!
8. I validated it with `signtool verify /pa sample.exe` and it worked: `0 sha256 Authenticode`

### Hacking electron-builder to use the right command
1. I found the existing command
a. I put in a fake `CSC_LINK` and `CSC_KEY_PASSWORD` and ran electron-builder. It errored and gave me a stack trace to `app-builder-lib`. I searched through that library and found `windowsCodeSign.js`, which creates the command.
2. I updated the command to match mine
a. Scrolled down to `computeSignToolArgs`
b. I deleted their file/password args with `args.splice(args.indexOf('/f'), 2);` and `args.splice(args.indexOf('/p'), 2);`
c. I added the sha1 arg with `args.push('/sha1')` and `args.push('THUMBPRINT-GOES-HERE')`

### Results
It works! ... It just asks me for my pin like 8 times, every time I make a build. 🤦‍♂️
